TableAdapter  & TableAdapter2x Vaneer adaption

rahulKQL opened the pull request:
This PR containes changes related to:
- Updated `TableAdapter` & `TableAdapter2x` to adapt to `v2.models.CreateTableRequest`.
- updated `ColumnDescriptorAdapter#buildGarbageCollectionRule` to adapt to  `v2.models.GCRules.GCRule`.
- `hbase1_x.BigtableAdmin`,  `hbase2_x.BigtableAdmin` & `AbstractBigtableAdmin` contains `toProto()` & **TOODs**, which will be removed with #1985.

sduskis commented on bigtable-hbase-2.x-parent/bigtable-hbase-2.x/src/main/java/com/google/cloud/bigtable/hbase2_x/adapters/admin/TableAdapter2x.java:
```diff
@@ -48,12 +48,12 @@
 
-  public static Table adapt(TableDescriptor desc) {
-    return adapt(new HTableDescriptor(desc), null).getTable();
+  public static CreateTableRequest adapt(TableDescriptor desc) {
```
Users should not be using this class.  They are more likely to use the Read adapters.  Please feel free to change this method.

sduskis commented on bigtable-hbase-2.x-parent/bigtable-hbase-2.x/src/main/java/com/google/cloud/bigtable/hbase2_x/adapters/admin/TableAdapter2x.java:
```diff
@@ -48,12 +48,12 @@
+    return adapt(new HTableDescriptor(desc), null);
   }
 
   public static ColumnFamily toColumnFamily(ColumnFamilyDescriptor column) {
```
Please feel free to remove this method.

sduskis commented:
We should tag all adapters with an InternalAPI annotation, if one exists in GAX.

rahulKQL commented:
- removed `TableAdapter2.x#adapt(TableDescripter)` & `TableAdapter2x#toColumnFamily` as it was not being used.
- added `@InternalApi` annotations on TableAdapter class.

sduskis commented on bigtable-client-core-parent/bigtable-hbase/src/main/java/com/google/cloud/bigtable/hbase/adapters/admin/TableAdapter.java:
```diff
@@ -30,6 +31,8 @@
 import java.util.Map;
 import java.util.Map.Entry;
 
+import static com.google.cloud.bigtable.hbase.adapters.admin.ColumnDescriptorAdapter.buildGarbageCollectionRule;
```
Please move the `import static` before the regular imports.  Also

sduskis commented on bigtable-client-core-parent/bigtable-hbase/src/main/java/com/google/cloud/bigtable/hbase/adapters/admin/TableAdapter.java:
```diff
@@ -41,37 +44,34 @@
    *
-   * @param desc a {@link org.apache.hadoop.hbase.HTableDescriptor} object.
-   * @return a {@link com.google.bigtable.admin.v2.Table} object.
+   * This method adapts ColumnFamily to CreateTableRequest.
```
Please use `{@link ...}` instead of using regular class names as strings.

sduskis commented on bigtable-client-core-parent/bigtable-hbase/src/main/java/com/google/cloud/bigtable/hbase/adapters/admin/TableAdapter.java:
```diff
@@ -41,37 +44,34 @@
    */
-  protected static Table adapt(HTableDescriptor desc) {
-    Map<String, ColumnFamily> columnFamilies = new HashMap<>();
+  @InternalApi
```
Please put the annotation at the class level.

rahulKQL replied:
I will update it to class level.

sduskis commented on bigtable-client-core-parent/bigtable-hbase/src/main/java/com/google/cloud/bigtable/hbase/adapters/admin/TableAdapter.java:
```diff
@@ -41,37 +44,34 @@
-    return builder;
+  @InternalApi
+  public static CreateTableRequest adapt(HTableDescriptor desc, byte[][] splitKeys) {
+    CreateTableRequest createTableRequest = CreateTableRequest.of(desc.getTableName().getNameAsString());
```
Please call the variable `request` for short

rahulKQL replied:
Would use shorter name here and for other occurrences as well.

sduskis commented on bigtable-client-core-parent/bigtable-hbase/src/main/java/org/apache/hadoop/hbase/client/AbstractBigtableAdmin.java:
```diff
@@ -121,6 +123,7 @@ public AbstractBigtableAdmin(CommonConnection connection) throws IOException {
     disabledTables = connection.getDisabledTables();
     bigtableInstanceName = options.getInstanceName();
     tableAdapter = new TableAdapter(bigtableInstanceName);
+    this.instanceName = InstanceName.of(bigtableInstanceName.getProjectId(), bigtableInstanceName.getInstanceId());
```
Please get the instance name from `bigtableInstanceName.toGcbInstanceName()`, if it exists.

rahulKQL replied:
Apparently `v2.models.CreateTableRequest` expects `com.gogle.bigtable.admin.v2.InstanceName`.
and `bigtableInstanceName.toGcbInstanceName()` returns `v2.models.InstanceName`.

igorbernstein2 replied:
Please note that I will be removing Names from the google-cloud-java to avoid these kind of headaches.
What is toGcbInstanceName?

sduskis replied:
@igorbernstein2, We have a `BigtableInstanceName` in this repo.  `toGcbInstanceName` converts the `BigtableInstanceName` to an `InstanceName`.  See [here](https://github.com/GoogleCloudPlatform/cloud-bigtable-client/commit/82b9bb6db4e861161f164cd5e514ec2d8f9e96c8#diff-36056d6784a87e2537c8ba1864c4d9f2R125).

sduskis commented on bigtable-client-core-parent/bigtable-hbase/src/main/java/org/apache/hadoop/hbase/client/AbstractBigtableAdmin.java:
```diff
@@ -354,13 +357,13 @@ public void createTable(HTableDescriptor desc, byte[] startKey, byte[] endKey, i
-  protected void createTable(TableName tableName, CreateTableRequest.Builder builder) throws IOException {
-    builder.setParent(bigtableInstanceName.toString());
+  //TODO(rahulkql):update methods to adapt to v2.models.CreateTableRequest
+  protected void createTable(TableName tableName, CreateTableRequest createTableRequest) throws IOException {
```
Please use the variable `request` instead of `createTableRequest`

sduskis commented on bigtable-client-core-parent/bigtable-hbase/src/main/java/org/apache/hadoop/hbase/client/AbstractBigtableAdmin.java:
```diff
@@ -370,14 +373,14 @@ protected void createTable(TableName tableName, CreateTableRequest.Builder build
     LOG.warn("Creating the table synchronously");
-    CreateTableRequest.Builder builder = TableAdapter.adapt(desc, splitKeys);
-    createTableAsync(builder, desc.getTableName());
+    CreateTableRequest createTableRequest = TableAdapter.adapt(desc, splitKeys).toProto(instanceName);
```
`request` variable name, please.

sduskis commented on bigtable-client-core-parent/bigtable-hbase/src/main/java/org/apache/hadoop/hbase/client/AbstractBigtableAdmin.java:
```diff
@@ -370,14 +373,14 @@ protected void createTable(TableName tableName, CreateTableRequest.Builder build
 
-  protected ListenableFuture<Table> createTableAsync(CreateTableRequest.Builder builder,
+  //TODO(rahulkql):update methods to adapt to v2.models.CreateTableRequest
+  protected ListenableFuture<Table> createTableAsync(CreateTableRequest createTableRequest,
```
`request` variable name, please.

sduskis commented on bigtable-hbase-2.x-parent/bigtable-hbase-2.x/src/main/java/com/google/cloud/bigtable/hbase2_x/BigtableAsyncAdmin.java:
```diff
@@ -112,7 +114,7 @@ public BigtableAsyncAdmin(CommonConnection asyncConnection) throws IOException {
     this.asyncConnection = asyncConnection;
     this.configuration = asyncConnection.getConfiguration();
-
+    this.instanceName = InstanceName.of(bigtableInstanceName.getProjectId(), bigtableInstanceName.getInstanceId());
```
Please use the `BigtableInstanceName.toGcbInstanceName()` here.

sduskis commented on bigtable-hbase-2.x-parent/bigtable-hbase-2.x/src/main/java/com/google/cloud/bigtable/hbase2_x/adapters/admin/TableAdapter2x.java:
```diff
@@ -79,6 +73,7 @@ public TableAdapter2x(BigtableOptions options) {
    * @param table a {@link com.google.bigtable.admin.v2.Table} object.
    * @return a {@link TableDescriptor} object.
    */
+  @InternalApi
```
Please use the `@InternalApi` at the class level.

sduskis commented:
@rahulKQL, it looks like the GCRule object has issues: https://github.com/googleapis/google-cloud-java/issues/4057.  Please add unit tests to TableAdapter.

igorbernstein2 commented on bigtable-client-core-parent/bigtable-hbase/src/main/java/com/google/cloud/bigtable/hbase/adapters/admin/ColumnDescriptorAdapter.java:
```diff
@@ -196,12 +199,12 @@ public static GcRule buildGarbageCollectionRule(HColumnDescriptor columnDescript
       // Intersection (AND)
       //    - maxAge = :HBase_ttlSeconds
       //    - maxVersions = :HBase_minVersion
-      ageRule = intersection(ageRule, maxVersions(minVersions));
```
Please remove the unused methods

rahulKQL replied:
sure, I will remove it

igorbernstein2 commented on bigtable-client-core-parent/bigtable-hbase/src/main/java/com/google/cloud/bigtable/hbase/adapters/admin/ColumnDescriptorAdapter.java:
```diff
@@ -181,12 +184,12 @@ public static GcRule buildGarbageCollectionRule(HColumnDescriptor columnDescript
 
     // minVersions only comes into play with a TTL:
-    GcRule ageRule = maxAge(ttlSeconds);
+    GCRule ageRule = GCRULES.maxAge(org.threeten.bp.Duration.ofSeconds(ttlSeconds));
```
Once you remove the unused methods, you should be able to remove the conflicting Duration import and import this one instead

igorbernstein2 commented on bigtable-client-core-parent/bigtable-hbase/src/main/java/com/google/cloud/bigtable/hbase/adapters/admin/TableAdapter.java:
```diff
@@ -41,37 +44,34 @@
-  protected static Table adapt(HTableDescriptor desc) {
-    Map<String, ColumnFamily> columnFamilies = new HashMap<>();
+  @InternalApi
+  protected static void adapt(CreateTableRequest createTableRequest, HTableDescriptor desc) {
```
Please move the output arg last so that it read adapt something input something else

rahulKQL commented:
- Added unit tests TestTableAdapter & TestTableAdapter2x.
- Removed unused methods from ColumnDescriptorAdapter.
- Changes for variable names & arguments

@sduskis  @igorbernstein2  
Could you please review it again.
